### ISSUE TITLE
Base rule extension: no-var configuration for declarations

### ISSUE THREAD

**Issue by DanKaplanSES** [typescript-eslint/typescript-eslint#7941]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+base+rule+extension%22) and found none that match my proposal.
- [X] I have searched the [current extension rule list](https://typescript-eslint.io/rules/#extension-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the base rule

https://eslint.org/docs/latest/rules/no-var

### Description

Typescript global declarations are more subtle than I ever expected. For a better understanding, I recommend these SO answers: 
* [Create a global variable in TypeScript](https://stackoverflow.com/a/56984941/61624) 
* [Using globalThis in Typescript](https://stackoverflow.com/a/68452689/61624)

Here are some key takeaways. If you want to add properties to `globalThis`:

1. When the typescript file has no `import` or `export` keyword (i.e., it's a script, not a module), it is done like so: `declare var age: number`.
2. When the typescript file does have `import` or `export`, you write this instead:
    ```ts
    // an import / export line

    declare global {
        var age: number;
    }
    ```

If `const` or `let` is used instead of `var`, `globalThis.age = 18` would error in both scenarios. So if the intention is to declare `globalThis.age`, it is *necessary* to use the `var` keyword: using `const` or `let` would be a mistake. When enabled, the base [no-var](https://eslint.org/docs/latest/rules/no-var) rule indiscriminately marks all `var` usage as wrong, but in this case, it is required.

### Fail

```typescript
/*eslint no-var: "error"*/
export {}

var x = "y";
var CONFIG = {};
```

### Pass

```typescript
export {}

declare global {
    var age: number;
}
```

**Comment by Josh-Cena** [comment]
See also: https://github.com/typescript-eslint/typescript-eslint/issues/7613, #2594

**Comment by bradzacher** [comment]
Honestly I think we're just fine not doing anything here - declaration files are a fraction of a fraction of a percent of almost every codebase - so building and maintaining *new* lint rules with exceptions is really not worth the effort or burden, IMO.

Better to document the small edge-case as good usecase for a disable comment and move on.

The big reason I think a disable is better is because it serves as clear documentation for future readers. Not everyone will know that a var must be used in that location - so commenting like below serves as an educational tombstone to prevent someone accidentally breaking it in future.

```js
eslint-diaable-next-line no-var -- we must use a var here so that the global declaration works as expected. Using let/const would cause incorrect behaviour.
```

**Comment by Josh-Cena** [comment]
> are a fraction of a fraction of a percent of almost every codebase

The _fraction_ is low, but their _existence_ is inevitable, though. Untyped dependencies, front end globals injected by scripts (like the Google API and FB API)...

**Comment by DanKaplanSES** [comment]
> See also: #7613, #2594
Huh. It's really strange that this didn't show up when I searched. Thanks for letting me know.

> Maybe this should be merged with https://github.com/typescript-eslint/typescript-eslint/issues/2594

I'm genuinely not sure if those failure cases are correct. Using `let` and `const` in a global type "works," it just works differently than a `var`. But is it bad do use `let` and `const` there?
 
@bradzacher 
> Better to document the small edge-case as good usecase for a disable comment and move on. 

Who is the persona in this comment, eslint maintainers or eslint users? 

> The big reason I think a disable is better...

Sorry, do you mean, "better than implementing this issue," "better than disabling the rule," "better than leaving it as a warning/error," or something else entirely?

**Comment by bradzacher** [comment]
> Sorry, do you mean ... something else entirely

I mean a disable comment, as I illustrated in the code block in my response.

> Do you mean eslint documentation will say this

Unlikely to get such a comment into ESLint core's docs - but likely fits as an FAQ article or even just these issues serve as documentation as they contain the relevant context.

---

@Josh-Cena for sure - they exist, though the vast majority of `.d.ts` files aren't dealing with global declarations. Again the vast, vast, vast majority of dependencies these days have types, and those that don't are modules that don't influence globals.

So we're talking case that impacts a fraction of a fraction of the codebase (just `.d.ts` files) for a fraction of a fraction of dependencies (just untyped dependencies that declare global types).

Most TS devs will never run into this and those that do will likely only touch such a file once and never again.

For sure it's a completely valid concern and something that we probably should handle! But it seems like something much better suited to "document and move on" rather than "implement an entire rule to handle"

**Comment by DanKaplanSES** [comment]
Thanks, @bradzacher I agree about the small number of eslint users this will affect. 

I'd like to document where this comes up for me:

1. A [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped) declaration file is out of sync with its implementation, and I have to troubleshoot the issue.
2. I'm porting JavaScript to TypeScript. 
3. I'm trying to write a browser library that can't use modules for whatever reason.
4. I'm trying to write an extension for a library that was written before modules or I'm trying to contribute to its code base.
5. I'm using JSDom and it's not pragmatic to isolate the `window` global from the node `global` environment. 

> it seems like something much better suited to "document and move on" rather than "implement an entire rule to handle"

I empathize and that works fine for me. It's others I worry about. They have to learn what I've learned before they can "document and move on," and it took me a while. But yeah, I get it: there are much better bang for your buck issues out there. :)

PS: This is besides the point, but this proposed rule applies to all typescript files, not just declaration files. You can put `declare global { ... }` in both. I prefer putting it in `.ts` files so the declaration and the implementation are nearby.

**Comment by bradzacher** [comment]
This was my personal opinion

> Most TS devs will never run into this and those that do will likely only touch such a file once and never again.
>
> For sure it's a completely valid concern and something that we probably should handle! But given we're already bandwidth constrained - it seems like something much better suited to "document and move on" rather than "implement an entire rule to handle"

And I have the same opinion for #2594 - it's 3 years old and has zero engagement (no reactions or comments). It's not an issue the community cares for so we should close it.

**Comment by Josh-Cena** [comment]
I think the rule is simple enough and has good educational impacts, so I'm not against implementing it. People would turn it on indirectly through their preset anyway—they don't have to be aware of its existence.

**Comment by bradzacher** [comment]
> and has good educational impacts

It wouldn't have any educational impact though - it would just stop erroring in this specific case. It wouldn't teach anyone anything!

**Comment by Josh-Cena** [comment]
The existence of this rule teaches people that they can and should use `var` in global definition files. We can add docs that further explain this. But I agree that the _actual_ education part has to come from the other rule proposed about enforcing global var instead of const.

**Comment by bradzacher** [comment]
We already have the base rule turned on in our `eslint-recommended` config.
If we add an extension rule we wouldn't create new reports here - we'd be removing reports.

So people don't know they have the rule turned on in the first place because it's turned on in a black-box extended config. And 99.999% of the time they don't get reports from the rule because people don't use `var` in modern TS. And in the rare case that someone does global augmentation they now don't get a report either.

So how are we improving education here? It seems to me like nobody would ever see a report or know the rule exists so nothing would change.

**Comment by Josh-Cena** [comment]
My point about "education" is not an error or a lack of error, which is why I agree a rule that enforces `var` is more educational because it gives you an error. My point is that:

1. The current state of erroring is counter-educational because people choose more awkward patterns that don't describe the runtime behavior faithfully, just to please the linter
2. The education comes from a documentation page. People would read it would understand why they need to relearn about `var` semantics in TS.

**Comment by oosawy** [comment]
I'd like to provide some context regarding how other toolchains are addressing this issue.

[Biome implemented this feature](https://github.com/biomejs/biome/pull/1669) earlier this year, while Deno lint, which supports TypeScript as a first-class citizen, [doesn't seem to have implemented it](https://github.com/denoland/deno_lint/blob/v0.31.0/src/rules/no_var.rs#L48) yet.

I can't estimate the actual cost of adding and maintaining a new rule, but seeing by the implementation of the no-var rule, it doesn't seem like it would be overly long or complex 10-20 lines even with this feature.

From an educational perspective, it would be beneficial to have a rule like prefer-declare-global-with-var as well. Many users inadvertently define global variables by using the interface Window {} merging. At least eslint's no-var rule, which implies that using var is incorrect, steers developers in the wrong way.

In fact, there's an instance where I recommended using var in a review, only for it to be changed to interface Window {} not sure if this rule's error triggered it or not.

**Comment by DanKaplanSES** [comment]
> If we wanted to create our own fork/extension of no-var...

Just using this quote as an opportunity to share a personal opinion: if this is implemented, it should be done in typescript-eslint and not an extension:

1. The extension would effectively _fix_ the overzealous `@typescript-eslint/no-var` rule which IMHO implies, "Sometimes `@typescript-eslint` steers you in the wrong direction." Optically, I don't like that message because it may cast doubt on other rules. 
    But for all I know, there are tons of extensions out there serving this purpose, in which case, I don't think I'm making a strong argument. 
1. `@typescript-eslint` and `eslint` have great documentation and extensions don't always live up to the same quality.
1. Extensions are harder to find (e.g., via Google). 
    1. It may not occur to a user that such an extension could exist.
1. From a user experience perspective, it's more convenient to have this built in than need to setup `@typescript-eslint` as a step then setup the extension as another step.

Apologies in advance if this is already obvious to everyone: all but the first apply to any extension, not just this one.

**Comment by kirkwaiblinger** [comment]
@DanKaplanSES 
to be clear, there is currently no `@typescript-eslint/no-var` rule. There is, externally, eslint's `no-var` rule, which is designed to work on JS. The question being considered in this issue is whether there are edge cases that occur when running that rule on TS code that are sufficiently misleading to warrant the creation of a `@typescript-eslint/no-var` rule to define TS-specific behavior. We'd refer to that as an "extension rule" or "base rule extension". When I said "fork" vs "extension", I was (somewhat confusingly) referring to the implementation details of the rule's code, which wouldn't have any user-facing impact (including on docs or anything else). The proposed extension would show up to the user just like, for example, require-await (https://typescript-eslint.io/rules/require-await/)

**Comment by DanKaplanSES** [comment]
> @DanKaplanSES to be clear, there is currently no `@typescript-eslint/no-var` rule. There is, externally, eslint's `no-var` rule, which is designed to work on JS. 

Ah, right. I had forgotten that since I first created this issue. Thanks for the reminder.

> The question being considered in this issue is whether there are edge cases that occur when running that rule on TS code that are sufficiently misleading to warrant the creation of a `@typescript-eslint/no-var` rule to define TS-specific behavior. We'd refer to that as an "extension rule" or "base rule extension". 

Ah, okay. In that case, in my previous reply, I'm preaching to the choir.

Thanks for clarifying!

**Comment by karlhorky** [comment]
Just dropping in here to mention that our students were also running into this problem now that we enabled the `no-namespace` rule, with both `no-var` reports on `declare global...var` and `no-namespace` reports on `declare module globalThis...let`:

- https://github.com/typescript-eslint/typescript-eslint/issues/9582

I proposed in the issue above that a blessed syntax be documented, but maybe it's better to instead change the `no-var` reporting - gut feeling is good on this option... 🤔

**Comment by JoshuaKGoldberg** [comment]
Coming over from #9865, I think I've landed in the camp of making a `@typescript-eslint/var` extension rule as @Josh-Cena and @kirkwaiblinger have suggested. The particularly convincing points for me were that:

* `declare global` contexts do pop up once in a while
* Having a dedicated docs page would be helpful for showing how to use a `declare global` `var` properly

**Comment by bradzacher** [comment]
My 2C:
In almost 12 months this issue has garnered just 4 community reactions.
It just doesn't seem like this is something that's important to the broader community.

**Comment by Josh-Cena** [comment]
It's not a bad practice developers are trying to lint out.
It's not widespread enough that adding a few eslint-disables is a nuisance.
It's even subtle enough that many just declare global consts and move on even when that doesn't reflect runtime reality.
All these mean we don't get strong incentive to vote on this, but that doesn't mean the rule is not good. Rather, exactly because of these points, it provides excellent educational opportunities if we can enforce the inverse and either use documentation or an explicit error to tell people to prefer var declarations.

**Comment by karlhorky** [comment]
> By-and-large the average TS user doesn't touch global declarations and the like.

Is this true? I would have thought that the global mutable var pattern would not be so niche - it's quite useful for simple global singletons, not sure I would classify this as a code smell or an antipattern.

**Comment by bradzacher** [comment]
With the advent of ESM -- yeah it is definitely niche and smelly!

Why would you create a global singleton via a global var when you can neatly export the singleton from a module and import it where required?
Augmenting `global` means:
- code analysis is hard -- you now have an implicit, untraceable dependency between modules.
    - EG bundlers cannot track where the global was defined and thus you could create a bundle without the required declaration spot -- creating a broken bundle.
- malicious actors can interact with your internals because it's readily available on `global`/`window`

Both of these are non-issues with ESM:
```ts
class MySingleton { ... }
export { type MySingleton };

const instance = new MySingleton();
export default instance;

// or perhaps you want a lazy init pattern
let instance: MySingleton | null = null;
export default function getOrInit(): MySingleton {
  if (instance == null) {
    instance = new MySingleton();
  }
  return instance;
}
```
- You can use tsconfig `paths` / bundler resolution config to create a path-less import if you want it to be simple like `import singleton from 'MySingleton';`
- This is immediately analysable by bundlers and tree-shakers alike.
- This singleton is not accessible outside your application closure.

**Comment by karlhorky** [comment]
Hmm, you're giving me something to consider 🤔 - maybe using an IIFE closure would be fine (at least for the first case that I'm thinking of now... not to say that I have thought of every possible use case):

### Before (with `globalThis`)

```ts
declare module globalThis {
  let postgresSqlClient: Sql;
}

// Connect only once to the database
// https://github.com/vercel/next.js/issues/7811#issuecomment-715259370
function connectOneTimeToDatabase() {
  if (!('postgresSqlClient' in globalThis)) {
    globalThis.postgresSqlClient = postgres(postgresConfig);
  }

  // Workaround to force Next.js Dynamic Rendering on every database query
  // https://github.com/vercel/next.js/discussions/50695
  return ((
    ...sqlParameters: Parameters<typeof globalThis.postgresSqlClient>
  ) => {
    noStore();
    return globalThis.postgresSqlClient(...sqlParameters);
  }) as typeof globalThis.postgresSqlClient;
}

export const sql = connectOneTimeToDatabase();
```

Source: https://github.com/upleveled/next-js-example-spring-2024-atvie/blob/34d5027a5670852b11c8812e0ac3c99cb0d933a4/database/connect.ts#L9-L41

### After (with local `let` variable in IIFE closure)

```ts
// Connect only once to the database
// https://github.com/vercel/next.js/issues/7811#issuecomment-715259370
const connectOneTimeToDatabase = (() => {
  let postgresSqlClient: Sql;

  return function connectOneTimeToDatabase() {
    if (!('postgresSqlClient' in globalThis)) {
      globalThis.postgresSqlClient = postgres(postgresConfig);
    }

    // Workaround to force Next.js Dynamic Rendering on every database query
    // https://github.com/vercel/next.js/discussions/50695
    return ((
      ...sqlParameters: Parameters<typeof globalThis.postgresSqlClient>
    ) => {
      noStore();
      return globalThis.postgresSqlClient(...sqlParameters);
    }) as typeof globalThis.postgresSqlClient;
  };
})();

export const sql = connectOneTimeToDatabase();
```

**Comment by bradzacher** [comment]
At the very least if you're going to use a global mutable variable for your DB client you should use a non-enumerable property with a module-private `Symbol` key for security so that someone can't grab your db client and mutate your db directly! Or worse - swap it out for their own client and steal your data!

**Comment by karlhorky** [comment]
I guess drifting a bit off topic here, but by "someone" I'm assuming you mean some code running in the Node.js global execution context? I guess if you have untrusted code running there you have bigger problems anyway and a symbol won't protect much.

**Comment by Josh-Cena** [comment]
Symbol properties do not grant you extra safety, because of `getOwnPropertySymbols`.

**Comment by bradzacher** [comment]
Wait what? Who specced those functions out to enumerate non-enumerable keys?!?
Thanks I hate it.

 But yes we are way off topic now!

**Comment by karlhorky** [comment]
> Hmm, you're giving me something to consider 🤔 - maybe using an IIFE closure would be fine for keeping a single database connection across hot reloading (at least for the first case that I'm thinking of now... not to say that I have thought of every possible use case of global mutable state)

So we looked into [the IIFE + closure option above](https://github.com/typescript-eslint/typescript-eslint/issues/7941#issuecomment-2309799234), and it looks like we still need `globalThis`:

1. React Fast Refresh (aka "hot reloading" or "live reloading") uses webpack Hot Module Replacement (React) or [Metro Hot Module Replacement (React Native 0.61+)](https://reactnative.dev/blog/2019/09/18/version-0.61) ([history](https://reactnative.dev/blog/2016/03/24/introducing-hot-reloading))
2. webpack and Metro HMR both re-initialize the top-level module scope when changes are made in a tree
3. This leads to [the `connectOneTimeToDatabase` constant](https://github.com/typescript-eslint/typescript-eslint/issues/7941#:~:text=const%20connectOneTimeToDatabase%20%3D%20(()%20%3D%3E%20%7B) being re-initialized, discarding the old closure and database connection / client
5. There appears to be [`module.hot.dispose(<cleanup fn>)` or `import.meta.webpackHot.dispose(<cleanup fn>)`](https://webpack.js.org/api/hot-module-replacement/#dispose-or-adddisposehandler), which would appear to be an option to end the database connection on every hot refresh, but callbacks passed to this are not called (potentially because of the automatic wrapping of React Fast Refresh)
6. Also, even if the `hot.dispose` option worked, it would be inefficient to always create and destroy new database connections on every hot reload

So it's my conclusion from all of this that `globalThis` is indeed the correct pattern for hot module replacement in dev server scenarios - not niche, a code smell or an antipattern.

Others also confirm this approach:

1. [Dev mode hot reload loses database connection · vercel/next.js · Discussion #26427](https://github.com/vercel/next.js/discussions/26427#discussioncomment-898062)
2. [How to preserve a database connection pool in hooks.server.js during hot reload in SvelteKit with Vite.js and mariadb? - Stack Overflow](https://stackoverflow.com/a/77167211/1268612)
3. etc.

This is of course only one use case for `globalThis` - gut feel says there are probably more non-smelly, non-antipattern, valid uses of it.

**Comment by bradzacher** [comment]
I dunno but to me it sounds like the global variable is easier, but is also hiding a bug in your dev env that could lead to unsafety or undefined behaviour.

You're spawning a db connection using one set of code. After a hot reload you have a new set of code that could be expecting to use a db connection that's setup with a different config.

So the only way to fix this state would be to hard reload your dev env to forcefully drop the connection and then recreate the entire connection using the new code.

**Comment by karlhorky** [comment]
This bug would only occur with changes to the database configuration:

- In dev environment
- Making a change which will trigger hot reloading
- Making a change to the database configuration

So, I would argue, more of an edge case. If this is a common pain point for developers (which I would doubt), then a comment at the configuration "Reload your dev server on any changes" would resolve it. Or if it's a common pain point and you want to allow developers to ignore comments: a wrapper script for the dev server to restart it on changes to the database config.

The main point here for this issue is that IMO this is a valid usage of mutable `globalThis`.

**Comment by bradzacher** [comment]
To be clear: I never said it wasn't valid to use `globalThis`.
I specifically said:

> By-and-large the average TS user doesn't touch global declarations and the like.

Which I still hold to be true.

In modern code generally it's a code smell to rely on mutable global variables and the vast, vast, vast majority of usecases, and in general people are instead going to rely solely on ESM to encapsulate things and declare dependencies explicitly.

That doesn't mean there aren't valid usecases for mutable variables in `globalThis` -- it's just that generally it is considered an anti-pattern or a code smell to use it in modern web because the alternatives are generally better.

**Comment by oosawy** [comment]
> By-and-large the average TS user doesn't touch global declarations and the like.

Exactly, global declarations is not everyday things, but there are 2.7k `declare global { var... }` and at least 10k+ `interface Window {}`. There may be some interpretations to the numbers, my concerns are the most of `interface Window {}` could be mistake as you know and caused by lack of knowledge or the no-var rule.
So it is not the linter's place to lead from a syntax that is well used and has valid use cases to a syntax that looks right but is wrong.

**Comment by NotWoods** [comment]
When I've worked with and taught other developers, there are sometimes points where they need to use a global variable. When they find `declare global` through searching and then see linter errors, what ends up happening is they assume they're on the wrong path. The end result is always to switch to `any` (`(window as any).globalFoo`) since they know it works.

I'd prefer the linter to default to supporting `declare global` without the need to disable rules, as it helps lead developers in the right direction.

**Comment by DanKaplanSES** [comment]
@NotWoods I think you bring up a separate issue. This one is about using `declare var` (within a `declare global`), not `declare global`.

**Comment by remcohaszing** [comment]
I think this issue can be closed in favor of https://github.com/eslint/eslint/pull/19714.
